app/upload.go

```diff
@@ -0,0 +1,214 @@
+	}
+
+	us.FileOffset = 0
+	now := time.Now()
+	us.CreateAt = model.GetMillisForTime(now)
+	us.Path = now.Format("20060102") + "/teams/noteam/channels/" + us.ChannelId + "/users/" + us.UserId + "/" + us.Id + "/" + us.Filename
+	if err := us.IsValid(); err != nil {
+		return nil, err
+	}
+
+	channel, err := a.GetChannel(us.ChannelId)
+	if err != nil {
+		return nil, model.NewAppError("CreateUploadSession", "app.upload.create.incorrect_channel_id.app_error",
+			map[string]interface{}{"channelId": us.ChannelId}, "", http.StatusBadRequest)
+	}
+	if channel.DeleteAt != 0 {
+		return nil, model.NewAppError("CreateUploadSession", "app.upload.create.cannot_upload_to_deleted_channel.app_error",
+			map[string]interface{}{"channelId": us.ChannelId}, "", http.StatusBadRequest)
+	}
+
+	us, storeErr := a.Srv().Store.UploadSession().Save(us)
+	if storeErr != nil {
+		return nil, model.NewAppError("CreateUploadSession", "app.upload.create.save.app_error", nil, storeErr.Error(), http.StatusInternalServerError)
+	}
+
+	return us, nil
+}
+
+func (a *App) GetUploadSession(uploadId string) (*model.UploadSession, *model.AppError) {
+	us, err := a.Srv().Store.UploadSession().Get(uploadId)
+	if err != nil {
+		var nfErr *store.ErrNotFound
+		switch {
+		case errors.As(err, &nfErr):
+			return nil, model.NewAppError("GetUpload", "app.upload.get.app_error",
+				nil, nfErr.Error(), http.StatusNotFound)
+		default:
+			return nil, model.NewAppError("GetUpload", "app.upload.get.app_error",
+				nil, err.Error(), http.StatusInternalServerError)
+		}
+	}
+	return us, nil
+}
+
+func (a *App) GetUploadSessionsForUser(userId string) ([]*model.UploadSession, *model.AppError) {
+	uss, err := a.Srv().Store.UploadSession().GetForUser(userId)
+	if err != nil {
+		return nil, model.NewAppError("GetUploadsForUser", "app.upload.get_for_user.app_error",
+			nil, err.Error(), http.StatusInternalServerError)
+	}
+	return uss, nil
+}
+
+func (a *App) UploadData(us *model.UploadSession, rd io.Reader) (*model.FileInfo, *model.AppError) {
+	// make sure it's not possible to upload more data than what is expected.
+	lr := &io.LimitedReader{
+		R: rd,
+		N: us.FileSize - us.FileOffset,
+	}
+	var err *model.AppError
+	var written int64
+	if us.FileOffset == 0 {
+		// new upload
+		written, err = a.WriteFile(lr, us.Path)
+		if written < minFirstPartSize && written != us.FileSize {
+			a.RemoveFile(us.Path)
+			return nil, model.NewAppError("UploadData", "app.upload.upload_data.first_part_too_small.app_error",
+				map[string]interface{}{"Size": minFirstPartSize}, "", http.StatusBadRequest)
+		}
+	} else if us.FileOffset < us.FileSize {
+		// resume upload
+		written, err = a.AppendFile(lr, us.Path)
+	}
+	if written > 0 {
+		us.FileOffset += written
+		if storeErr := a.Srv().Store.UploadSession().Update(us); storeErr != nil {
+			return nil, model.NewAppError("UploadData", "app.upload.upload_data.update.app_error", nil, storeErr.Error(), http.StatusInternalServerError)
```

Comment:
@mgdelacroix 

The only problem I can see with this is that we end up making an additional store call to reset the session which could fail as well (since the original failed it's quite possible) and cause the same issue. I am not sure it's going to be easy to completely avoid all possible inconsistencies with the current design.

This could be potentially improved by *truncating* the stored file instead of removing everything. That way it's like the update never happened but it involves a bit more work on the `FileStore` side.

Reply:
Good point, I totally agree. I'd say that we can move forward with the current implementation and if it ever causes problems implement the file truncation in the store, I don't think this is a blocker